### DISCUSSION TITLE
Make prometheus go install-able

### DISCUSSION THREAD

**Discussion by Syphdias** [prometheus/prometheus#12611] [Ideas]
### Proposal

I would like to be able to build prometheus including the web asset files with `go generate` and  `go build`. This would enable two things:
1. Installation via `go install`
2. Using upstream prometheus with https://github.com/gokrazy/gokrazy/ instead of building it locally first

There are two issues that already talked about this in the past (2015: https://github.com/prometheus/prometheus/issues/548 and 2020: https://github.com/prometheus/prometheus/issues/7517). The latter mentions a mailing list post where someone explains that "the new React UI requires some separate build steps that are not Go-based, and the result of that is not checked in (because it's hard to make completely reproducible builds that wouldn't create diffs every time the React UI is built)." It then continues to link to where the magic happens: a dead link and lines in the Makefile but on master instead of a commit.

My hope is that the React UI is no longer a limiting factor to making prometheus work with `go install` (formerly `go get`).

**Comment by roidelapluie** [comment]
This is not the limiting factor. The limiting factor is the "replace" in go.mod.

**Comment by stapelberg** [comment]
> This is not the limiting factor. The limiting factor is the "replace" in go.mod.

Can you elaborate on that please? Both `go install` and gokrazy’s tooling should work just fine with replace directives AFAIK.

**Comment by roidelapluie** [comment]
```
$ go install github.com/prometheus/prometheus/cmd/prometheus@latest
go: github.com/prometheus/prometheus/cmd/prometheus@latest (in github.com/prometheus/prometheus@v0.46.0):
        The go.mod file for the module providing named packages contains one or
        more replace directives. It must not contain directives that would cause
        it to be interpreted differently than if it were the main module.
```

**Comment by stapelberg** [comment]
TIL, thanks. In case others are wondering: https://github.com/golang/go/issues/40276 → “Why can't module replace directives be used?”. I’ll see if I can raise this with the team in case this can be made work.

Either way, gokrazy does not run into this limitation, because it prepares a build directory and then runs `go build`. I checked:

```
% gok -i prom new
% gok -i prom add github.com/prometheus/prometheus/cmd/prometheus@latest
% gok -i prom overwrite --root /tmp/root.squashfs
% unsquashfs -ll /tmp/root.squashfs | grep prom
-rwxr-xr-x root/root         116916591 2023-07-26 08:51 squashfs-root/user/prometheus                                

```

**Comment by roidelapluie** [comment]
Then you can download the UI:

https://github.com/prometheus/prometheus/releases/download/v2.46.0/prometheus-web-ui-2.46.0.tar.gz

and place it in (I think ./web/ui), then build with `builtinassets,stringlabels` tags

**Comment by Syphdias** [comment]
Thank you, @stapelberg, for taking a closer look. I am out of my depth here.

I probably misunderstood the problem why the web assets were not included when using prometheus with gokrazy. I thought gokrazy would do a `go install`, I also thought this would imply `go generate` and `go build`. Both of theses  assumptions are probably wrong.
